### DISCUSSION TITLE
Use zephyr Settings API with AT24C256 EEPROM as backend

### DISCUSSION THREAD

**Discussion by MaazSk** [zephyrproject-rtos/zephyr#79243] [Q&A]
Hello Everyone,
I want to use zephyr settings API with AT24C256 EEPROM as a backend. I am not sure how to do the same. Is there any example to do it specifically a eeprom as a backend. I went through the example in samples/subsys/settings but was unable to find out how to add EEPROM as a backend any example would be great. I am using STM32H750B_DK as the dev board

**Comment by Laczen** [comment]
There is no in tree support for using settings on eeprom.

**Reply by MaazSk** [reply]
Can we add custom backend as eeprom?

**Reply by Laczen** [reply]
What do you mean? Can you contribute a eeprom settings backend? Of course.

**Reply by MaazSk** [reply]
> Can you contribute a eeprom settings backend? Of course.

This is something I can try. 

But this is what I am trying and what I wanted to do is, use EEPROM as a custom backend to store the settings. I came across [custom backend](https://docs.zephyrproject.org/latest/services/settings/index.html#example-custom-backend-implementation) where we can use a custom backend to store settings. I am curious to know if I can use EEPROM as a backend using these custom backend api's?

**Reply by Laczen** [reply]
The CUSTOM_BACKEND kconfig option is to signal the settings config that a alternative backend will be provided. There is no custom backend API. When a settings backend for eeprom is contributed it could be selected by EEPROM_BACKEND. It needs to provide a settings store solution and a `settings_backend_init`.

**Reply by MaazSk** [reply]
So can we have the eeprom calls inside `settings_custom_load` and `settings_custom_save` just want to know is it possible or not?

Like follows

```
static int settings_custom_load(struct settings_store *cs,
                                const struct settings_load_arg *arg)
{
    eeprom_read(eeprom_dev, address, &value);
}

static int settings_custom_save(struct settings_store *cs, const char *name,
                                const char *value, size_t val_len)
{
    eeprom_read(eeprom_dev, address, value);
}
```

**Reply by Laczen** [reply]
Yes, you can use `eeprom_read` and `eeprom_write` inside these routines. But you need to create some kind of storage solution on eeprom that can read/write key(name)-value data. The `settings_custom_load()` will need to walk over the data in the eeprom and call the `settings_call_set_handler()` for each of the key(name)-value pairs that are on the eeprom. The `settings_custom_save()` will need to save the key(name)-value combination to eeprom in such a way that it can be retrieved by `custom_settings_load()`.

**Reply by mirkobeckerPxC** [reply]
I'm also very interested in this topic. Is there anything new on this?

**Reply by kpochwala** [reply]
I also stumbled upon this. I am using ESP32S3 with external 8M PSRAM, and it turns out that it is [not possible to write into internal Flash while using external PSRAM](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-reference/peripherals/spi_flash/spi_flash_concurrency.html) - I'm getting cache errors which seems like the problem from linked article. I want to use the settings subsystem for storage of [WiFi credentials](https://docs.zephyrproject.org/latest/services/connectivity/networking/api/wifi_credentials.html), I can easily bodge-wire an i2c eeprom to the current board, and if it would do the trick, then the next batch could have the eeprom on board. Is anybody currently working on this?
